UPGRADE_GUIDE.md

```diff
@@ -2,22 +2,130 @@
+This change allows for multiple clients to be instantiated with different API keys and for the library to be used in a
+multi-threaded environment.
+
+```java
+// Old
+EasyPost.apiKey = System.getenv("EASYPOST_API_KEY"); // global API key
+
+// New
+EasyPostClient client1 = new EasyPostClient("EASYPOST_API_KEY_1"); // per-client API key
+EasyPostClient client2 = new EasyPostClient("EASYPOST_API_KEY_2"); // per-client API key
+```
+
+All functions are now accessed via the `EasyPostClient` instance. For example, to create a shipment:
+
+```java
+// Old
+EasyPost.apiKey = System.getenv("EASYPOST_API_KEY");
+Shipment shipment = Shipment.create(shipmentMap);
+
+// New
+EasyPostClient client = new EasyPostClient("EASYPOST_API_KEY_1");
+Shipment shipment = client.shipment.create(shipmentMap);
+```
+
+## 6.0 Error Types
+
+*Likelihood of Impact: **High***
+
+Error handling has been overhauled and a number of specific exception types have been introduced.
+
+All exceptions inherit from `EasyPost.Exception.EasyPostException` (which extends `System.Exception`). Users can catch this exception type to handle all errors thrown by the library.
+
+Subclasses of `EasyPostException` are grouped into two categories:
+
+- `EasyPost.Exception.API` for errors returned by the API. Subclasses of this exception type are:
+  - `EasyPost.Exception.ExternalApiError` - thrown when an issue occurs with an external API (e.g. Stripe)
+  - `EasyPost.Exception.ForbiddenError` - thrown when you don't have permission to access this API resource
+  - `EasyPost.Exception.GatewayTimeoutError` - thrown when the API gateway times out (504 status code)
+  - `EasyPost.Exception.InternalServerError` - thrown when an internal server error occurs (500 status code)
+  - `EasyPost.Exception.InvalidRequestError` - thrown when the API received an invalid request (422 status code)
+  - `EasyPost.Exception.MethodNotAllowedError` - thrown when the API receives a request with an invalid HTTP method (
+        405 status code)
+  - `EasyPost.Exception.NotFoundError` - thrown when the API receives a request for a resource that does not exist (
+        404 status code)
+  - `EasyPost.Exception.PaymentError` - thrown when a payment error occurs (402 status code)
+  - `EasyPost.Exception.ProxyError` - thrown when the library is unable to connect to the API via a proxy
+  - `EasyPost.Exception.RateLimitError` - thrown when the API rate limit is exceeded (429 status code)
+  - `EasyPost.Exception.RedirectError` - thrown when the http status is between 300 and 308.
+  - `EasyPost.Exception.ServiceUnavailableError` - thrown when the API is unavailable (503 status code)
+  - `EasyPost.Exception.TimeoutError` - thrown when the API request times out (408 status code)
+  - `EasyPost.Exception.UnauthorizedError` - thrown when the API receives an unauthorized request (401 or 403 status
+        code)
+  - `EasyPost.Exception.UnknownApiError` - thrown when an unknown API error occurs (unexpected 4xx status code)
```

Comment:
Is the name space `Exception` or `Exceptions`? Inconsistent here, double-check throughout.

Reply:
Exception in Java